cpp/src/slice2cs/CsUtil.cpp

```diff
@@ -941,47 +961,31 @@ Slice::CsGenerator::writeTaggedMarshalCode(Output &out,
     {
         const TypePtr elementType = seq->type();
         const bool isArray = !seq->hasMetaDataWithPrefix("cs:generic:");
-        const string length = isArray ? param + ".Length" : param + ".Count";
         builtin = BuiltinPtr::dynamicCast(elementType);
-        st = StructPtr::dynamicCast(elementType);
 
-        if(elementType->isVariableLength())
+        if (seq->hasMetaDataWithPrefix("cs:serializable:"))
         {
-            out << nl << "if (" << param << " != null && " << stream << ".WriteOptional(" << tag << ", "
-                << getTagFormat(seq, scope) << "))";
-            out << sb;
-            out << nl << "var pos = " << stream << ".StartSize();";
-            writeMarshalCode(out, seq, scope, param, stream);
-            out << nl << stream << ".EndSize(pos);";
-            out << eb;
+            out << nl << stream << ".WriteSerializable(" << tag << ", " << param << ");";
         }
-        else if(st)
+        else if (elementType->isVariableLength())
         {
-            // Fixed size struct
-            out << nl << "if (" << param << " != null && " << stream << ".WriteOptional(" << tag << ", "
-                << getTagFormat(seq, scope) << "))";
-            out << sb;
-            if(st->minWireSize() > 1)
-            {
-                out << nl << stream << ".WriteSize(" << length << " * " << st->minWireSize() << " + (" << length
-                    << " > 254 ? 5 : 1));";
-            }
-            writeMarshalCode(out, seq, scope, param, stream);
-            out << eb;
+            out << nl << stream << ".WriteTaggedSeq(" << tag << ", " << param << ", " <<
+                outputStreamWriter(elementType, scope, true) << ");";
         }
-        else if(seq->hasMetaDataWithPrefix("cs:serializable:"))
+        else if (builtin && isArray && builtin->isNumericTypeOrBool())
         {
-            assert(builtin->kind() == Builtin::KindByte);
-            out << nl << "if (" << param << " != null && " << stream << ".WriteOptional(" << tag
-                << ", " << getUnqualified("Ice.OptionalFormat", scope) << ".VSize))";
-            out << sb;
-            out << nl << stream << ".WriteSerializable(" << param << ");";
-            out << eb;
+            out << nl << stream << ".Write" << builtinSuffixTable[builtin->kind()] << "Seq(" << tag << ", " << param;
+            if (!isDataMember)
+            {
+                out << ".Span";
+            }
```

Comment:
Not so easy, because only sequence of fixed-size numeric without metadata use `.Span`, and the caller of this function doesn't check the types of the params.